File: 0901-1000/0979-Distribute Coins in Binary Tree/0979-Distribute Coins in Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def distributeCoins(self, root: TreeNode) -> int:
        total = 0
        def dfs(root):
            if root is None:
                return 0

            left = dfs(root.left)
            right = dfs(root.right)
            nonlocal total
            total += abs(left) + abs(right)
            return root.val - 1 + left + right

        dfs(root)
        return total

File: 0901-1000/0979-Distribute Coins in Binary Tree/0979-Distribute Coins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        int total = 0;
        dfs(root, total);
        return total;
    }

private:
    int dfs(TreeNode* root, int& total) {
        if (root == nullptr) {
            return 0;
        }

        int left = dfs(root->left, total);
        int right = dfs(root->right, total);
        total += abs(left) + abs(right);
        return root->val - 1 + left + right;
    }
};
